solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java

```diff
@@ -266,13 +271,14 @@ public void downloadConfig(String configName, Path dir) throws IOException {
   @Override
   public List<String> getAllConfigFiles(String configName) throws IOException {
     String zkPath = CONFIGS_ZKNODE + "/" + configName;
-    if (!zkPath.startsWith(ZkConfigSetService.CONFIGS_ZKNODE + "/")) {
-      throw new IllegalArgumentException("\"" + zkPath + "\" not recognized as a configset path");
-    }
     try {
-      List<String> filePaths = new LinkedList<>();
-      for (String filePath : zkClient.getAllConfigsetFiles(zkPath)) {
-        filePaths.add(filePath.replaceAll(zkPath + "/", ""));
+      List<String> filePaths = new ArrayList<>();
+      ZkMaintenanceUtils.traverseZkTree(zkClient, zkPath, ZkMaintenanceUtils.VISIT_ORDER.VISIT_POST, filePaths::add);
+      filePaths.remove(zkPath);
+      for (int i=0; i<filePaths.size(); i++) {
+        String filePath = filePaths.get(i);
+        assert(filePath.startsWith(zkPath + "/"));
+        filePaths.set(i, filePath.replaceAll(zkPath + "/", ""));
```

Comment:
Don't use String.replaceAll when you are looking for something at the beginning or end.  replaceAll has to do more work than is needed and looks for other occurrences when it shouldn't have to here.  I know calling replaceAll is "easy" on the caller (you) but in Solr we pay more attention to performance.

Given that assert statement, it seems you are confident the results start with zkPath; yes?  (tests pass?).  I think traverseZkTree ought to not return paths that start with is parameter; it seems wrong to me.

Reply:
yes, the results start with zkPath (=/configs/configName)
traverseZkTree includes this zkPath (I debugged)